pkg/request/pool.go

```diff
@@ -216,39 +234,231 @@ func (rp *Pool) NextRequests(ctx context.Context) [][]byte {
 	return rawRequests
 }
 
-func (rp *Pool) RemoveRequests(requests ...string) error {
-	if atomic.LoadUint32(&rp.batchingEnabled) == 0 {
-		rp.pending.RemoveRequests(requests...)
-		return nil
+func (rp *Pool) RemoveRequests(requestsIDs ...string) {
+	rp.lock.RLock()
+	defer rp.lock.RUnlock()
+
+	if !rp.isBatchingEnabled() {
+		rp.pending.RemoveRequests(requestsIDs...)
+		return
 	}
 
-	for _, requestID := range requests {
+	for _, requestID := range requestsIDs {
 		rp.batchStore.Remove(requestID)
 	}
-	return nil
+	return
 }
 
-// Close removes all the requests, stops all the timeout timers.
+func (rp *Pool) Prune(predicate func([]byte) error) {
+	rp.lock.RLock()
```

Comment:
I understand the generic implementation here fits well to the pattern, however in practice it's extremely inefficient to enumerate all requests serially like that. 

Can we just implement an efficient parallel `Prune()` in both the pending and batch store and just invoke it accordingly? The pending can prune each bucket in parallel and the batch store can just iterate over the sealed batches in parallel.

Reply:
ok I'll find a way to make this faster
although this prune is called only after a verification sequence change, so not that often...

Reply:
Yeah you're right, up to you then.

Reply:
Will save this for later then